### ISSUE TITLE
💅 noConfusingVoidType false positive

### ISSUE THREAD

**Issue by steelbrain** [biomejs/biome#1812]
### Environment information

```bash
CLI:
  Version:                      1.5.3
  Color support:                true

Platform:
  CPU Architecture:             aarch64
  OS:                           macos

Environment:
  BIOME_LOG_DIR:                unset
  NO_COLOR:                     unset
  TERM:                         "xterm-256color"
  JS_RUNTIME_VERSION:           "v18.19.0"
  JS_RUNTIME_NAME:              "node"
  NODE_PACKAGE_MANAGER:         "yarn/3.6.4"

Biome Configuration:
  Status:                       Loaded successfully
  Formatter disabled:           false
  Linter disabled:              false
  Organize imports disabled:    true
  VCS disabled:                 false

Workspace:
  Open Documents:               0
```

### Rule name

noConfusingVoidType

### Playground link

https://biomejs.dev/playground/?code=dAB5AHAAZQAgAEMAcgBlAGEAdABlAE0AbwBkAGUAcgBhAHQAZQBkAEUAbgB0AGkAdAB5AFIAZQBzAHUAbAB0ADwAVAA%2BACAAPQAKACAAIAB8ACAAKAAoAFQAIABlAHgAdABlAG4AZABzACAAdgBvAGkAZAAgAD8AIABSAGUAYwBvAHIAZAA8AHMAdAByAGkAbgBnACwAIABuAGUAdgBlAHIAPgAgADoAIABUACkAIAAmACAAewAKACAAIAAgACAAIAAgAHMAdABhAHQAdQBzADoAIAAnAHMAdQBjAGMAZQBzAHMAJwAKACAAIAAgACAAfQApAAoAIAAgAHwAIAB7AAoAIAAgACAAIAAgACAAcwB0AGEAdAB1AHMAOgAgACcAYgBsAG8AYwBrAGUAZAAnAAoAIAAgACAAIAAgACAAbQBlAHMAcwBhAGcAZQA6ACAAcwB0AHIAaQBuAGcACgAgACAAIAAgAH0ACgAgACAAfAAgAHsACgAgACAAIAAgACAAIABzAHQAYQB0AHUAcwA6ACAAJwBzAHUAcwBwAGUAbgBkAGUAZAAnAAoAIAAgACAAIAB9AAoACgB0AHkAcABlACAAQQAgAD0AIABDAHIAZQBhAHQAZQBNAG8AZABlAHIAYQB0AGUAZABFAG4AdABpAHQAeQBSAGUAcwB1AGwAdAA8AHYAbwBpAGQAPgAKAHQAeQBwAGUAIABCACAAPQAgAEMAcgBlAGEAdABlAE0AbwBkAGUAcgBhAHQAZQBkAEUAbgB0AGkAdAB5AFIAZQBzAHUAbAB0ADwAewAgAHAAbwBzAHQASQBkADoAIABuAHUAbQBiAGUAcgAgAHwAIABuAHUAbABsACAAfQA%2BAA%3D%3D

### Expected result

`something extends void` is a valid way to infer the generic type information and provide conditioned results, it should be allowed.

### Code of Conduct

- [X] I agree to follow Biome's Code of Conduct
